### PR TITLE
chore: remove duplicate code from dash and simplify

### PR DESCRIPTION
**Pull request by kostasrim**
* rename all Policy members for consistency 
* remove duplicate code

### CODE REVIEW

**Comment by dranikpg** [comment]
Yay my favourite refactoring PRs 🥳🥳

File: src/core/dash_internal.h [hunk 1]
```diff
@@ -337,32 +341,27 @@ template <typename _Key, typename _Value, typename Policy = DefaultSegmentPolicy
       std::swap(value[slot_a], value[slot_b]);
     }
 
-    // calls for each busy slot: cb(iterator, probe)
-    template <typename Cb> void ForEachSlot(Cb&& cb) const {
+    template <typename This, typename Cb> void ForEachSlotImpl(This obj, Cb&& cb) const {
```
Comment (romange) [inline src/core/dash_internal.h:344]: 1. curious, why you added This template? 
2. You also use `this` at line 345-346. is it intended?

File: src/core/dash_internal.h [hunk 2]
```diff
@@ -337,32 +341,27 @@ template <typename _Key, typename _Value, typename Policy = DefaultSegmentPolicy
...
 
-      for (unsigned j = 0; j < NUM_SLOTS; ++j) {
-        if (mask & 1) {
-          cb(this, j, probe_mask & 1);
-        }
-        mask >>= 1;
-        probe_mask >>= 1;
-      }
+    // calls for each busy slot: cb(iterator, probe)
+    template <typename Cb> void ForEachSlot(Cb&& cb) {
```
Comment (romange) [inline src/core/dash_internal.h:363]: I understand now

File: src/core/dash_internal.h [hunk 3]
```diff
@@ -1462,33 +1456,31 @@ std::enable_if_t<UV, unsigned> Segment<Key, Value, Policy>::CVCOnInsert(uint64_t
...
+    if (bucket_[bid].GetVersion() < ver_threshold)
+      bid_res[cnt++] = bid;
 
-    if (!bucket_[after_next].IsEmpty() && bucket_[after_next].GetVersion() < ver_threshold)
-      bid_res[cnt++] = after_next;
+    if (!bucket_[nid].IsEmpty() && bucket_[nid].GetVersion() < ver_threshold)
+      bid_res[cnt++] = nid;
+  };
 
+  if (CheckIfMovesToOther(true, nid, after_next)) {
```
Comment (romange) [inline src/core/dash_internal.h:1470]: is there a functional change?

File: src/core/dash_internal.h [hunk 4]
```diff
@@ -1546,34 +1538,27 @@ std::enable_if_t<UV, unsigned> Segment<Key, Value, Policy>::CVCOnBump(uint64_t v
...
-  unsigned stash_pos = bid - kRegularBucketCnt;
+  unsigned stash_pos = bid - kBucketNum;
   uint8_t fp_hash = hash & kFpMask;
 
   auto find_stash = [&](unsigned, unsigned pos) {
     return stash_pos == pos ? slot : BucketType::kNanSlot;
   };
 
-  if (target.GetVersion() < ver_threshold) {
+  auto do_fun = [&result, &result_bid, fp_hash, find_stash](auto& target, auto target_bid,
```
Comment (romange) [inline src/core/dash_internal.h:1548]: is there a functional change?

File: src/core/dash_internal.h [hunk 5]
```diff
@@ -1636,11 +1621,8 @@ bool Segment<Key, Value, Policy>::TraverseLogicalBucket(uint8_t bid, HashFn&& hf
 
 template <typename Key, typename Value, typename Policy>
 size_t Segment<Key, Value, Policy>::SlowSize() const {
-  size_t res = 0;
-  for (unsigned i = 0; i < kTotalBuckets; ++i) {
-    res += bucket_[i].Size();
-  }
-  return res;
+  return std::accumulate(&bucket_[0], &bucket_[0] + kTotalBuckets, size_t(0),
```
Comment (romange) [inline src/core/dash_internal.h:1624]: I would like to revert this change :) just because `std::acccumulate` exists, it does not mean using it makes the code more readable

**Review by romange — COMMENTED** [review]
thanks for improving the code here!

File: src/core/dash_internal.h (hunk 5 above)
Comment (dranikpg) [reply src/core/dash_internal.h:1624]: I agree, accumulate has some "mental" reading overhead... Really anticipate ranges until it becomes view(bucket_, bucket_ + offset).map(size).sum()

Comment (kostasrim) [reply src/core/dash_internal.h:1624]: IMO in this particular case I think accumulate is fine but I am more than happy to revert! @dranikpg I wanted to make the exact same comment but you got it first. Note that:

1. ranges::accumulate is not part of c++20 unfortunately but I think it's available in c++23 as `left_fold`. And you can replace the custom lambda with a `projection` (similar to `.map` you wrote)
2. boost::accumulate would suffice here as well

Generally, my main complain against STL algorithms (as with most of the world) is that you have to pass the range explicitly. I would personally opt in (as suggested in the past as well) to use one of the two libraries in our codebase and make our life a little easier with using those algorithms.

**Review by dranikpg — APPROVED** [review]
